src/descriptor/mod.rs

```diff
@@ -848,7 +978,153 @@ impl<Pk: MiniscriptKey> fmt::Display for Descriptor<Pk> {
             Descriptor::Sh(ref sub) => write!(f, "sh({})", sub),
             Descriptor::Wsh(ref sub) => write!(f, "wsh({})", sub),
             Descriptor::ShWsh(ref sub) => write!(f, "sh(wsh({}))", sub),
+            Descriptor::ShSortedMulti(ref smv) => write!(f, "sh({})", smv),
+            Descriptor::WshSortedMulti(ref smv) => write!(f, "wsh({})", smv),
+            Descriptor::ShWshSortedMulti(ref smv) => write!(f, "sh(wsh({}))", smv),
+        }
+    }
+}
+
+/// Contents of a "sortedmulti" descriptor
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SortedMultiVec<Pk: MiniscriptKey, Ctx: ScriptContext> {
+    /// signatures required
+    pub k: usize,
+    /// public keys inside sorted Multi
+    pub pks: Vec<Pk>,
+    /// The current ScriptContext for sortedmulti
+    pub(crate) phantom: PhantomData<Ctx>,
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn new(k: usize, pks: &[Pk]) -> Result<Self, Error> {
+        // Check the limits before creating a new SortedMultiVec
+        // For example, under p2sh context the scriptlen can only be
+        // upto 520 bytes.
+        let term: miniscript::decode::Terminal<Pk, Ctx> = Terminal::Multi(k, Vec::from(pks));
+        let ms = Miniscript::from_ast(term)?;
+
+        // This would check all the consensus rules for p2sh/p2wsh and
+        // even tapscript in future
+        Ctx::check_ms_validity(&ms)?;
+
+        Ok(Self {
+            k,
+            pks: Vec::from(pks),
+            phantom: PhantomData,
+        })
+    }
+    /// Parse an expression tree into a SortedMultiVec
+    fn from_tree(tree: &Tree) -> Result<Self, Error>
+    where
+        <Pk as FromStr>::Err: ToString,
+    {
+        let k = tree.args[0].name.parse::<usize>().unwrap(); // FIXME: unwrap
+        let pks: Result<Vec<Pk>, _> = tree.args[1..]
+            .iter()
+            .map(|sub| expression::terminal(sub, Pk::from_str))
+            .collect();
+        pks.map(|pks| SortedMultiVec::new(k, &pks))?
+    }
+}
+
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    fn translate_pk<FPk, Q, FuncError>(
+        &self,
+        translatefpk: &mut FPk,
+    ) -> Result<SortedMultiVec<Q, Ctx>, FuncError>
+    where
+        FPk: FnMut(&Pk) -> Result<Q, FuncError>,
+        Q: MiniscriptKey,
+    {
+        let pks: Result<Vec<Q>, _> = self.pks.iter().map(&mut *translatefpk).collect();
+        Ok(SortedMultiVec {
+            k: self.k,
+            pks: pks?,
+            phantom: PhantomData,
+        })
+    }
+}
+
+impl<Pk: MiniscriptKey + ToPublicKey, Ctx: ScriptContext> SortedMultiVec<Pk, Ctx> {
+    /// Create Terminal::Multi containing sorted pubkeys
+    pub fn sorted_node(&self) -> Terminal<bitcoin::PublicKey, Ctx> {
```

Comment:
We should change this to return `Terminal<Pk, Ctx>` while still sorting using the internal public keys.
`        pks.sort_by(|a, b| a.to_public_key().partial_cmp(&b.to_public_key()).unwrap())`

Reply:
Thanks for help on these trait bounds. This `sort_by` statement will compare the 64-byte representation of the pubkeys from secp256k1, not lexigraphic representation as required by BIP 67. Calling `to_pubkey().to_string()`  does do lexigraphic comparison, however.

Reply:
Ah, good catch.

Reply:
We can add comment about bip 67 comparison here.

Reply:
Is the `Sort pubkeys lexigraphically according to BIP 67` comment enough?

Reply:
We should fix the `Ord` impl upstream to not sort on the internal representation, because that's crazy and has burned multiple projects.

But in the meantime we can use `.encode` rather than `to_string` and avoid an allocation.

Reply:
Also we can use `sort_by_key` rather than `sort_by`.

I'd prefer `sort_by_cached_key` but unfortunately we don't have that with our MSRV..

Reply:
I meant `serialize` rather than `encode`, which returns a `[u8; 33]`, and because Rust is stupid we'll have to further wrap that in a type because it doesn't come with its own `Ord` impl.

So my new proposal is to define, within this module,
```
pub struct Bip67Sorted<'a, Pk: ToPublicKey>(&'a Pk);
impl<'a, Pk: ToPublicKey> Ord for Bip67Sorted<'a, Pk> {
    fn cmp(&self, other: &Self) -> cmp::Ordering {
        (&self.to_public_key().serialize()[..]).cmp(&other.to_public_key().serialize()[..])
    }
}
impl<'a, Pk: ToPublicKey> PartialOrd for Bip67Sorted<'a, Pk> {
    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
        Some(self.cmp(other))
    }   
}
```
with suitable modifications to make that compile since I just typed it into github without testing :).

We should also move it to the top-level `lib.rs`, unless Sanket can think of a better place for it. I'd like to expose this since I want to use it in Liquid :P.

Reply:
Responded here: https://github.com/rust-bitcoin/rust-miniscript/pull/171/files/079955beb9c469b0c5c1aae43d3bfd0061176807#r520250242